src/cloud-search-index-storage.js

```diff
@@ -0,0 +1,210 @@
+   * from the provided object that are specified to be used in the index and saves a record with
+   * those values; other values are ignored.
+   * @param {Object} doc the object to store in the index.
+   * @returns An object containing the saved record's ObjectID.
+   */
+  async save(doc) {
+    const indexRecord = Object.keys(doc).reduce((obj, key) => {
+      const result = clone(obj);
+      if (this.#FIELDS_TO_INDEX.has(key)) {
+        result[key] = doc[key];
+      }
+      return result;
+    }, {});
+    if (!indexRecord?.objectID) {
+      indexRecord.objectID = crypto.randomUUID();
+    }
+
+    this.log.info(`Saving doc to cloud record storage index ${this.indexName}`, indexRecord);
+    return this.index.saveObject(indexRecord);
+  }
+
+  /**
+   * Get a list of ObjectID that contain the contentHash
+   * @param {String} contentHash sha256 hash of the source binary
+   * @returns list of strings of ojectIDs
+   */
+  async getObjectIdsByContentHash(contentHash) {
+    const searchResult = await this.index.search(
+      '',
+      {
+        distinct: false,
+        facetFilters: [
+          `contentHash:${contentHash}`,
+        ],
+      },
+    );
+    if (searchResult?.nbHits > 0) {
+      return searchResult?.hits.map((hit) => hit.objectID);
+    }
+    return false;
+  }
+
+  /**
+   * Update all records in the index containing the contentHash
+   *
+   * - Only update the values that are provided in the object (partial update)
+   * - Only add fields that are able to be indexed: `this.#FIELDS_TO_INDEX`
+   * @param {Object} doc document containing the contentHash and other fields to update
+   * @returns
+   */
+  async updateByContentHash(doc) {
+    if (!doc?.contentHash) {
+      this.log.info('Missing contentHash, no update in the record storage.');
+      return undefined;
+    }
+    let objectIDs = await this.getObjectIdsByContentHash(doc.contentHash);
+    if (!objectIDs) {
+      this.log.info('No matching record found in the record storage.');
+      return undefined;
+    }
+    objectIDs = objectIDs.map((objectID) => ({ objectID, ...doc }));
+    return this.update(objectIDs);
+  }
+
+  /**
+   * Update records in a index
+   * Only update the values that are provided in the object (partial update):
+   * https://www.algolia.com/doc/api-reference/api-methods/partial-update-objects/
+   * @param {Array<Object>} records Array of content records with fields to update
+   * @returns
+   */
+  async update(records) {
+    const indexRecords = records.map((doc) => {
+      const indexRecord = Object.keys(doc).reduce((obj, key) => {
+        const result = clone(obj);
+        if (this.#FIELDS_TO_INDEX.has(key)) {
+          result[key] = doc[key];
+        }
+        return result;
+      }, {});
+      return indexRecord;
+    });
+    this.log.info(`Updating docs in cloud record storage index ${this.indexName}`, indexRecords);
```

Comment:
Would it make sense to log the number of documents you are updating? Actually, to go a step further:

Debug = Log the full array
Info = Log the count